### ISSUE TITLE
Video becomes tiny after coming out of fullscreen

### ISSUE THREAD

**Issue by SumYuanLoong** [xternal7/ultrawidify#24]
Browser: FF 58.0.2 (64-bit)
Using latest version 2.1.2 on AMO
On youtube, after cycling in and out of full screen, the video is reduced to a tiny box on top center.
![firefox_2018-02-22_00-36-01](https://user-images.githubusercontent.com/18534656/36492913-47833f90-1769-11e8-923c-aa334a6dbe7d.png)

have not seen this issue on embedded videos on other sites.
possible that YT pushed an update that cause the issue as I have not updated anything related for the past week and issue only appeared since yesterday.

Fullscreen mode is not affected, issue is only with theater mode in the new youtube

**Comment by xternal7** [comment]
I'll take it #23 is trying to report this same issue, then, but with zero info. 

Can reproduce, with fresh install as well. I'll be looking into this over the weekend (hopefully I'll get some time). Until I manage to fix this, workaround is to use either full screen or normal (non-theater) mode I guess.

Interestingly enough it only affects Firefox, while Chrome doesn't seem to have this issue.

**Comment by delano8** [comment]
I had the same problem, however not even with fullscreen but just randomly about 50% of the time (I do always use cinema mode), and was super annoyed with version 2+ anyway because autodetect doesn't always work and the shortcut buttons don't always work either for some reason. Anyway, I didn't feel like waiting for a fix, so I went back to 1.2.1.1, because that works perfectly for what I want (youtube only, and I can click to fix ratio instead of remembering stuff).

So, cool, everything works again. And then I went back to reddit. And I realized, the problem I've been having on reddit, where basically all gifs were broken for about a week now (gifs auto close, disappear etc etc), were also fixed.

Soooo, somehow version 2.1.2, which really messed up my youtube experience, ALSO really messed up my reddit experience. And I never even though it could be this addon, but it sure was. I'm really happy to be back with 1.2.1.1 and hope I can stay with it forever, it's just simply better and more stable for me.

Don't fix it if it isn't broken, right?

(Firefox 58.0.2 win10 64 bit)

**Comment by TangoFox4** [comment]
Haven the issue too. It happens when I play a video in theatre mode and then drag or resize the window.

**Comment by xternal7** [comment]
@delano8 

>Don't fix it if it isn't broken, right?

#8 #5 (there's two or three reviews on AMO alone that mirrored #8), and the code was generally unmaintainable due to shitty coding practices of a CS student who's never really written javascript stuff before this extension.

And yes, GUI buttons were a huge part of it (and they could break the embedded youtube player on G+) so they aren't going to come back in the foreseeable future - at least not in their past form. There's the extension button next to the tab bar now. 

![image](https://user-images.githubusercontent.com/12505802/36530194-bc2cc43c-17ba-11e8-8fb0-0ed7fe7535de.png)

(Also you can disable extension per-site in the topmost tab of that popup). 

**I'm not completely ruling out a different implementation of in-player GUI, though.** The piece of code that got borked by a recent youtube update gives me the rectangle in which video gets displayed. That means I can (after making workaround for youtube) calculate where to put the element with control buttons. At the moment, the priorities for this extension are:

0. fix youtube screwy and blacklist reddit by default (should have done that earlier).
1. finish improvements to autodetection (probably not the bug @delano8 mentioned back there, I think I fixed that about a week ago?) and handle porting of extension settings between versions. Some people had some issues where extension broke until reinstalled, and corrupted settings seemed to be the problem.
2. reintroduce gradual zoom on `z` and `u` and provide a way to 'scroll' the zoomed in video up/down left/right
3. reintroduce settings page (rebindable keys, blacklist/whitelist management, some settings for automatic aspect ratio detection)
4. site-specific options for sites that require additional CSS classes or other hacks (see: vimeo, which is disable)
5. figure the best way to do GUI (injecting buttons into the player bar is _not_ a good way. Been there, done that, each site has its own way and some appear to be impossible), but motivation is lacking because jesus fuck some people are fucking salty and toxic about that. They're legit more salty than me after major bamboozle by one of my D&D groups.
6. Stretch mode, because some people are very salty and toxic about the fact that this extension is here to solve a problem that's different than the one they want. More salty than me rn.


>it would probably be good if you would put a support email address somewhere, so people don't have to create gifthub accounts just to report issues.

Suppose that's a valid point. I'll consider it (strongly) for the next non-bugfix release. Apparently `mailto` links allow you to set subject and body, so I'm prolly try to automate some data collection that I have to ask for way too often along the way.

**Comment by delano8** [comment]
I understand the button problems. But for me, and I think a lot of people, 1.2.1.1 works fine.
-I only use it on youtube.
-I don't have to remember anything.
-It doesn't break any other sites like reddit.
-It just simply works, without any hassle (for me).
For me that's way more then enough. I don't need an auto detect feature it it doesn't work 100% of the time (for instance when the 21:9 black bars on top and bottom are not perfectly black, but a more greyish tint, it doesn't seem to work).
I personally don't need all those features that can break, I just want a simple working button, even one button would be enough, the first one in the row, it's all I've ever used.

**Comment by xternal7** [comment]
>For me that's way more then enough. I don't need an auto detect feature it it doesn't work 100% of the time (for instance when the 21:9 black bars on top and bottom are not perfectly black, but a more greyish tint, it doesn't seem to work).

And I thought I was a bit too generous on the threshold of how dark the black bars need to be. Do you know any videos that do this off the top of your head? That would help with testing.

**Comment by TangoFox4** [comment]
To defend the 2.0, I love it. It's a bless when it just works without the need of dooing something. I think it's a change to the right direction.

**Comment by xternal7** [comment]
Thanks, much appreciated.

**Comment by xternal7** [comment]
Okay then, so apparently github autocloses issues based on commit message content. Neat.

Anyway, I kinda rushed this fix through and didn't check whether other sites are broken. The fix appears to work just fine. 

If extension doesn't work following the update (which I'm uploading to AMO as I type this), please tell me and then reinstall the extension. Fresh install verified works.

**Comment by delano8** [comment]
Clean install. Reddit gifs no longer broken, and no tiny screen after resizing atm. I still personally liked the button as well, but I guess autodetect is easier in the long run. Thanks for fixing! I'll get back to you on the grey bars when I come to it again.

**Comment by xternal7** [comment]
Thanks for the videos.

>first video

Oof. Black is #0f0f0f (treshold is #0A0A0A). I guess the 'detect the darkest color and assume it's black' stuff is going to get written.

>edit2

I can't verify that because I broke development version, but I'll go ahead and assume that aspect ratio of that video is just right for automatic detection to fail. 

Automatic detection picks 7 rows and 7 columns (i think, every n/8 except the edges). On the intersection of said rows and columns it picks a pixel and looks at it color.

After picking the pixels, it does a quick and dirty test of aspect ratio. First it checks in which of those 7 rows the black bars end and the non-black material begins. So if the first row is all black but the second row has a few pixels that aren't black -> letterbox starts somewhere between first and second row. 

Then, it performs the same check at the bottom. Is there black pixels in the bottom-most block and a few non-black pixels in the second-bottom-most block? Letterbox starts somewhere between first and second block from the bottom.

Because we don't want to correct letterbox that isn't letterbox, this extension assumes that letterbox must be centered. There is some tolerance that allows video to be slightly off-center, but that tolerance isn't taken into account for the first, quick and dirty stage.

Of course, that means that under some very specific conditions, a video could be only a pixel off-center (or not even that, depending on what rows get sampled. My math could be off by a pixel or two.), but the extension would assume the video is not centered because the bottom letterbox doesn't equal to the top one, thus doing nothing.

And that's what I referenced with "improvements to autodetection" there in my numbered list.

>For me, the beginning of the video suffers some zoom flashing from the autodetect.

Apparently fix for #12 wasn't enough. The cause seems to be that the frame has dark uneven top and bottom edge that gets approximately centered. That's a bit harder issue to crack, but hopefully I'll manage to do it in the next autodetection update. There's this bit where I'm checking if the first colored pixel in the column is on a horizontal line of pixels.

Basically, assume numbers are non-black pixels while zero are black pixels. We search for where the line between black and non-black pixels begin. Eventually, we find it.

```
2548670984729087309874928374324
1231234341237×432341445400123
0000000000000_0000000000000000
000000000000000000000000000000
```
So we know that × is first non-black pixel. We know that _ is the last black pixel. If we look to the left and to the right of either, we expect to see only black neighbours for _ and mostly non-black neighbours for ×. We check if our expectations are correct. If either of the rules is violated (namely, if pixels in the line of slightly left and right of _ aren't all black, or if pixels left and right of × are mostly black), we aren't detecting the letterbox, we are probably detecting an edge in the image.

Another fix that'll eventually come in is that if autodetection detects too many aspect ratio changes in too short time (e.g. flashing), it'll just delay changing aspect ratio for a few frames.
